### ISSUE TITLE
Up and down arrow keys jerks too quickly + patch

### ISSUE THREAD

**Issue by RobertBColton** [atom/tree-view#1165]
### Description

On the tree view if you use the down arrow keys to go all the way down and then slowly press the up key a couple of times to come back up the whole scroll area will jerk up too far instead of just far enough to show the item.

This right here is the problem:
https://github.com/atom/tree-view/blob/8d28d89ca6de7ef442bff000236ea33ca76379ca/lib/tree-view.coffee#L772

That line should actually pass false because we do not want the up/down arrow keys to center the item in the view. We just want it to scroll far enough to make the new selection visible (what we get in VS Code). VS Code does this without jerking that badly.

The reason VS Code does not have this is because it uses a custom variant of scrollIntoViewIfNeeded that only adjusts the most immediate parent and does not center the item.
https://github.com/Microsoft/vscode/blob/3f043b2da69a78cb19bed3e59823714be81afd48/src/vs/base/parts/tree/browser/treeDefaults.ts#L232

VS Code does the same for general list widgets too:
https://github.com/Microsoft/vscode/blob/a188171bf8d74e3ac1b54c0cd18469cdd471e442/src/vs/base/browser/ui/list/listWidget.ts#L264

This is the VS Code ``reveal()`` code which does basically what scrollIntoViewIfNeeded does but without jerking:
https://github.com/Microsoft/vscode/blob/a188171bf8d74e3ac1b54c0cd18469cdd471e442/src/vs/base/browser/ui/list/listWidget.ts#L821

### Steps to Reproduce

1. Open a project that shows the vertical scroll bar on the tree view
2. Hold the down arrow key to select the very last item in the tree
3. Now just press the up key enough times to select the third item up that's out of the view.

**Expected behavior:** [What you expect to happen]
The view should scroll up just about 24px or so, just enough to cover the selected item that's out of view.

**Actual behavior:** [What actually happens]
When this item gets selected the scroll will center it (undesirable, it scrolls too far up).

**Reproduces how often:** [What percentage of the time does it reproduce?]
This occurs always.

### Versions

I am on Windows 10.
```bash
$ atom --version

Atom    : 1.19.2
Electron: 1.6.9
Chrome  : 56.0.2924.87
Node    : 7.4.0

```

**Comment by 50Wliu** [comment]
Thanks, I can reproduce this.  And a double-thanks for doing the investigation and figuring out the root cause 🙇.  Are you interested in creating a PR to fix this?

**Comment by RobertBColton** [comment]
Sure, but I am confused because it looks like in my local install for 1.19.2, the tree-view files are all JavaScript. This repo seems to be CoffeeScript, so I am not exactly sure how to test the fix.

**Comment by 50Wliu** [comment]
Here's the Flight Manual article to get you started: http://flight-manual.atom.io/hacking-atom/sections/contributing-to-official-atom-packages/.  It even uses Tree View as the example! :smile:

If you have any questions, feel free to keep asking here or on [Slack](https://atom-slack.herokuapp.com) where I'm `@wliu`.

**Comment by RobertBColton** [comment]
Ok, I managed to get as far as apm install, but that's throwing syntax errors:
```
C:\Users\Owner\AppData\Local\atom\app-1.19.2\resources\app\apm\node_modules\.bin\node-gyp:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^
SyntaxError: missing ) after argument list
```

**Comment by 50Wliu** [comment]
Can you try again using Powershell instead of a bash-like prompt?

**Comment by RobertBColton** [comment]
Sure, here is using PowerShell from VS Code's integrated terminal:
```
Windows PowerShell
Copyright (C) 2016 Microsoft Corporation. All rights reserved.

PS C:\Users\Owner\Desktop\Projects\tree-view> apm install
Installing modules failed
> pathwatcher@8.0.0 install C:\Users\Owner\Desktop\Projects\tree-view\node_modules\pathwatcher
> node-gyp rebuild


C:\Users\Owner\Desktop\Projects\tree-view\node_modules\pathwatcher>if not defined npm_config_node_gyp (node "C:\Users\Owner\AppData\Local\atom\app-1.19.2\resources\app\ap
m\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Users\Owner\AppData\Local\atom\app-1.19.2\resources\app\apm\bi
n\\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
MSBUILD : error MSB3428: Could not load the Visual C++ component "VCBuild.exe". To fix this, 1) install the .NET Framework 2.0 SDK, 2) install Microsoft Visual Studio 200
5 or 3) add the location of the component to the system path if it is installed elsewhere.  [C:\Users\Owner\Desktop\Projects\tree-view\node_modules\pathwatcher\build\bind
ing.sln]

gyp ERR! build error
gyp ERR! stack Error: `C:\Windows\Microsoft.NET\Framework\v4.0.30319\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Users\Owner\AppData\Local\atom\app-1.19.2\resources\app\apm\node_modules\node-gyp\lib\build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\Users\\Owner\\AppData\\Local\\atom\\app-1.19.2\\resources\\app\\apm\\bin\\node.exe" "C:\\Users\\Owner\\AppData\\Local\\atom\\app-1.19.2\\resources\\
app\\apm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\Owner\Desktop\Projects\tree-view\node_modules\pathwatcher
gyp ERR! node -v v6.9.5
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
npm ERR! Windows_NT 10.0.15063
npm ERR! argv "C:\\Users\\Owner\\AppData\\Local\\atom\\app-1.19.2\\resources\\app\\apm\\bin\\node.exe" "C:\\Users\\Owner\\AppData\\Local\\atom\\app-1.19.2\\resources\\app
\\apm\\node_modules\\npm\\bin\\npm-cli.js" "--globalconfig" "C:\\Users\\Owner\\.atom\\.apm\\.apmrc" "--userconfig" "C:\\Users\\Owner\\.atom\\.apmrc" "install" "--runtime=
electron" "--target=1.6.9" "--arch=x64"
npm ERR! node v6.9.5
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE

npm ERR! pathwatcher@8.0.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pathwatcher@8.0.0 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the pathwatcher package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs pathwatcher
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls pathwatcher
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Owner\Desktop\Projects\tree-view\npm-debug.log
```

Command prompt gives the same result unfortunately.

**Comment by 50Wliu** [comment]
Hmm...maybe we're missing some info on that page.  Can you try installing the requirements at https://github.com/atom/atom/blob/master/docs/build-instructions/windows.md#requirements?

**Comment by RobertBColton** [comment]
Thanks, that did the trick. I had just uninstalled VS 2015 to install VS 2017 and didn't want to reinstall 2015. So I just downloaded the Visual C++ 2015 Build Tools instead.

Once I was able to successfully run ``apm install`` I created pull request #1171.

**Comment by 50Wliu** [comment]
Fixed by #1171.

**Comment by RobertBColton** [comment]
Could you also please take a look at atom/flight-manual.atom.io#369 because I really feel that one should be merged.
